### PR TITLE
Feat testing

### PR DESCRIPTION
**Pull request by lucarin91**

### CODE REVIEW

File: kvs_test.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+	t.Run("already exist", func(t *testing.T) {
+		kvs := NewMyKVS()
+
+		kvs.Store("SjknVAAA", "golang.org")
+		err := kvs.Store("SjknVAAA", "golang.org")
+		if nil == err {
+			t.Fail()
+		}
+		if "already exist" != err.Error() {
```
Comment (arl) [inline kvs_test.go:38]: Testing for a specific error string by comparing the message returned by `err.Error()` is not advised since the error message could change. When you do want to test against a specific error and not an error _in general_, then it's preferable to test against the error type or error variable (sentinel errors).

Resources: more [here](https://blog.golang.org/go1.13-errors)

TLDR: One way to _solve_ this:
Adding a `var errAlreadyExists errors.New("can't add store entry: already exists")` global, returning it when and comparing against it in tests.

File: router_test.go [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+
+		req := httptest.NewRequest("GET", "/", nil)
+		w := httptest.NewRecorder()
+		s.redirect(w, req)
+
+		resp := w.Result()
+		body, _ := ioutil.ReadAll(resp.Body)
+
+		if http.StatusBadRequest != resp.StatusCode {
+			t.Fail()
```
Comment (arl) [inline router_test.go:23]: same as before: you probably should use t.Error instead of t.Fail and add some context.

File: router_test.go [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+		resp := w.Result()
+		body, _ := ioutil.ReadAll(resp.Body)
+
+		if http.StatusOK != resp.StatusCode {
+			t.Fail()
+		}
+		r, _ := json.MarshalIndent(s.Stats, "", "  ")
+		if string(r)+"\n" != string(body) {
+			t.Fail()
+		}
```
Comment (arl) [inline router_test.go:186]: That's one way to do it. There's another where you wouldn't need to take care of adding `\n` though.
It would be preferable if you had decoded the response body from JSON to a `Stats` struct.
By doing so you'd just have to compare if the decoded struct is equal to `s.Stats`. 
The stdlib function for this comparison is [reflect.DeepEqual](https://pkg.go.dev/reflect#DeepEqual)

File: stats_test.go [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+		t.Fail()
+	}
+
+	for i, v := range s.Handlers {
+		if newS.Handlers[i].Name != v.Name ||
+			newS.Handlers[i].Count != v.Count {
+			t.Fail()
+		}
+	}
+}
```
Comment (arl) [inline stats_test.go:32]: Same as before, simpler with `reflect.DeepEqual`

File: stats_test.go [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+	s := Stats{
+		ServerStats{
+			1,
+			Redirects{
+				2,
+				3,
+			},
+			[]Handler{Handler{"qwe", 4}, Handler{"qwee", 5}, Handler{"qweee", 6}},
+		},
+	}
```
Comment (arl) [inline stats_test.go:17]: In general it's preferable to use the name of the fields in literal structs declarations.
Why? If for whatever reason the order of the fields was modified in the struct definition : 
 - at best you'd have a compile error, for example the first field was an int and is now a slice
 - at worst the code still compiles because types are respected but now you're not assigning the wrong values to the fields!

Example:

```go
type BankTransaction struct {
    Cents int
    Dollars int 
}
```

to 

```go
type BankTransaction struct {
    Dollars int 
    Cents int
}
```
:D

**Review by arl — CHANGES_REQUESTED** [review]
Nice, a very good job :+1: 
Most of the comments I let are about Go conventions, idioms or stuff you haven't discover yet.
